Framework/Algorithms/src/FitPeaks.cpp

```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+  getRangeData(histogram, range, vec_x, vec_y, vec_e);
+  // sum up all counts
+  double total = std::accumulate(vec_y.begin(), vec_y.end(), 0.);
   return total;
 }
 
 //----------------------------------------------------------------------------------------------
-/** Get number of vector elements falling into specified range
- * @param vector :: vector of type T
+/** Estimate background noise given a peak-window Y-vector
+ * @param vec_y :: a peak window of Y-values
+ * @return :: noise estimate
+ */
+double estimateBackgroundNoise(const std::vector<double> &vec_y) {
+  // assume that the first and the last 10 data points are background. Combine them into a data vector.
+  assert(vec_y.size() > 20);
+  std::vector<double> vec_bkg;
+  vec_bkg.resize(20);
+  std::copy(vec_y.begin(), vec_y.begin() + 10, vec_bkg.begin());
+  std::copy(vec_y.end() - 10, vec_y.end(), back_inserter(vec_bkg));
+  // estimate the noise as the standard deviation of that combined data vector
+  auto intensityStatistics = Kernel::getStatistics(vec_bkg, StatOptions::CorrectedStdDev);
+  return intensityStatistics.standard_deviation;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Calculate signal-to-noise ratio for a histogram range
+ * @param histogram :: histogram instance
+ * @param range :: histogram range
+ * @return :: signal-to-noise ratio
+ */
+double calculateSignalToNoiseRatio(const Histogram &histogram, const std::pair<double, double> &range,
+                                   const API::IBackgroundFunction_sptr &bkgd_function) {
+  // convert X range boundaries to index boundaries
+  size_t left_index, right_index;
+  const auto &orig_x = histogram.x();
+  rangeToIndexBounds(orig_x, range.first, range.second, left_index, right_index);
+
+  // handle an invalid range case
+  if (left_index >= right_index)
+    return 0.0;
+
+  // estimate background level by Y(X) fitting
+  if (!estimateBackgroundParameters(histogram, std::pair<size_t, size_t>(left_index, right_index), bkgd_function))
+    return 0.0; // failed to estimate background parameters
+
+  // get X,Y,and E for the data range
+  std::vector<double> vec_x, vec_y, vec_e;
+  getRangeData(histogram, range, vec_x, vec_y, vec_e);
+  if (vec_x.empty())
+    return 0.0;
+
+  // reduce all Y-values in the data range by the background
+  reduceByBackground(bkgd_function, vec_x, vec_y);
+
+  // estimate the signal as the highest Y-value in the data range
+  auto it_max = std::max_element(vec_y.begin(), vec_y.end());
+  double signal = vec_y[it_max - vec_y.begin()];
+  if (signal <= DBL_MIN)
+    return 0.0;
+
+  // estimate noise from background. Normal experimental data should always have some background noise.
+  // Zero noise usually indicates a problem with the experiment. Note that artificial data generated by
+  // FitPeaksTest might also have zero background noise.
+  double noise = estimateBackgroundNoise(vec_y);
+  if (noise <= DBL_MIN)
```

Comment:
`PDCalibration` (which calls this algorithm`FitPeaks` is used at ISIS to fit peaks in a spectrum of a single detector pixel that often contains lots of zero bins  - depending on the background region chosen the standard deviation may well be zero - it looks like in this instance the noise would be set to `DBL_MAX` and therefore $signal/noise \gtrapprox 0$ - whereas in this case if the peak bins have any intensity we might want to try fitting this peak?

Reply:
In this case the returned DBL_MAX is actually a signal-to-noise value, not a noise value.  In other words, peaks like this will automatically pass the signal-to-noise test. Having said that, I have removed the comment referring to "zero noise" as something irregular and added a comment about the DBL_MAX.

Reply:
Thanks!